(unfinished) runtime-sdk: gas

pro-wh opened the pull request:
fixes #89

the story so far:

- setting up a new `tx_values` field in TxContext similar to `values`, but it's transaction-local
- there's no batch-wide gas limit
- the core module manages a `core.GasUsed` tx_value with a u64 of how much gas the tx has used
- it provides a `use_gas` API function to increase that and check it against the tx auth info's fee's gas limit
- unit tests for that
- adding the skeleton for a `core.EstimateGas` query method, the idea from #89. dunno how to set up a new context inside a query yet tho

nhynes commented on runtime-sdk/src/context.rs:
```diff
@@ -177,6 +178,9 @@ pub struct TxContext<'a, 'b> {
     values: &'b mut BTreeMap<&'static str, Box<dyn Any>>,
+
+    /// Per-transaction values.
+    tx_values: BTreeMap<&'static str, Box<dyn Any>>,
```
should this also be borrowed as `&'b mut`?

pro-wh replied:
this one is owned, with the lifecycle matching the context itself

nhynes commented on runtime-sdk/src/context.rs:
```diff
@@ -271,6 +275,31 @@ impl<'a, 'b> TxContext<'a, 'b> {
     }
+
+    /// Fetches or sets a value associated with the transaction.
+    pub fn tx_value<V>(&mut self, key: &'static str) -> &mut V
```
```suggestion
    pub fn tx_value<V>(&mut self, key: &str) -> &mut V
```
I think that the static requirement is a bit heavy, but w/e

pro-wh replied:
could this apply to existing `values` related APIs too? we could open a separate issue if the scope is thus larger

kostko replied:
Yes this would apply to the existing APIs as well. No strong preference, but in practice these keys will have static lifetime as they will be defined as constants.

pro-wh replied:
thx. moving to #104

nhynes commented on runtime-sdk/src/context.rs:
```diff
@@ -271,6 +275,31 @@ impl<'a, 'b> TxContext<'a, 'b> {
+    {
+        self.tx_values
+            .entry(key)
+            .or_insert_with(|| Box::new(V::default()))
```
```suggestion
            .or_insert_with(Default::default)
```
should work

kostko replied:
Oh, in this case feel free to also fix the existing stuff.

pro-wh replied:
ok

pro-wh replied:
Oh you know what, it's inserting into `...Map<..., Box<dyn Any>>`, so all it knows is the `Any` in this context, so it doesn't know to use `V`

```
error[E0277]: the trait bound `(dyn std::any::Any + 'static): std::default::Default` is not satisfied
   --> runtime-sdk/src/context.rs:109:9
    |
109 | /         self.values
110 | |             .entry(key)
111 | |             .or_insert_with(Default::default)
    | |_____________________________________________^ the trait `std::default::Default` is not implemented for `(dyn std::any::Any + 'static)`
    | 
   ::: /home/wh0/.rustup/toolchains/nightly-2020-08-29-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/default.rs:84:1
    |
84  |   pub trait Default: Sized {
    |   ------------------------ required by this bound in `std::default::Default::default`
    |
    = note: required because of the requirements on the impl of `std::default::Default` for `std::boxed::Box<(dyn std::any::Any + 'static)>`

error[E0277]: the trait bound `dyn std::any::Any: std::default::Default` is not satisfied
   --> runtime-sdk/src/context.rs:111:29
    |
111 |             .or_insert_with(Default::default)
    |                             ^^^^^^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `dyn std::any::Any`
    |
    = note: required because of the requirements on the impl of `std::default::Default` for `std::boxed::Box<dyn std::any::Any>`
    = note: required by `std::default::Default::default`
```

nhynes commented on runtime-sdk/src/context.rs:
```diff
@@ -271,6 +275,31 @@ impl<'a, 'b> TxContext<'a, 'b> {
+            .entry(key)
+            .or_insert_with(|| Box::new(V::default()))
+            .downcast_mut()
+            .expect("type should stay the same")
```
I generally like to use `unwrap()` with a comment to prevent useless strings from filling up my `.text` section

kostko replied:
No strong preference, just make it consistent also in existing code.

pro-wh replied:
do we get good stack traces?

pro-wh replied:
this seems to be prevalent in places this PR won't :crossed_fingers: touch. let's split this out into a separate issue

pro-wh replied:
https://github.com/oasisprotocol/oasis-sdk/issues/99

nhynes commented on runtime-sdk/src/modules/core/test.rs:
```diff
@@ -0,0 +1,36 @@
+};
+
+#[test]
+fn test_use_gas() {
```
```suggestion
fn test_use_gas() {
    const MAX_GAS = 1000;
```
this constant is suggested for use below to make the code clearer due to fewer magic numbers

nhynes commented on runtime-sdk/src/modules/core/test.rs:
```diff
@@ -0,0 +1,36 @@
+    let mut mock = mock::Mock::default();
+    let mut ctx = mock.create_ctx();
+    let mut tx = mock::transaction();
+    tx.auth_info.fee.gas = 1000;
```
```suggestion
    tx.auth_info.fee.gas = MAX_GAS;
```

nhynes commented on runtime-sdk/src/modules/core/test.rs:
```diff
@@ -0,0 +1,36 @@
+    tx.auth_info.fee.gas = 1000;
+
+    ctx.with_tx(tx.clone(), |mut tx_ctx, _call| {
+        Core::use_gas(&mut tx_ctx, 100)
```
```suggestion
        Core::use_gas(&mut tx_ctx, MAX_GAS)
```

nhynes commented on runtime-sdk/src/modules/core/test.rs:
```diff
@@ -0,0 +1,36 @@
+    });
+
+    ctx.with_tx(tx.clone(), |mut tx_ctx, _call| {
+        Core::use_gas(&mut tx_ctx, 950)
```
```suggestion
        Core::use_gas(&mut tx_ctx, MAX_GAS)
```

nhynes commented on runtime-sdk/src/modules/core/test.rs:
```diff
@@ -0,0 +1,36 @@
+    });
+
+    ctx.with_tx(tx.clone(), |mut tx_ctx, _call| {
+        Core::use_gas(&mut tx_ctx, 600).unwrap();
```
```suggestion
        Core::use_gas(&mut tx_ctx, MAX_GAS).unwrap();
```

nhynes commented on runtime-sdk/src/modules/core/test.rs:
```diff
@@ -0,0 +1,36 @@
+
+    ctx.with_tx(tx.clone(), |mut tx_ctx, _call| {
+        Core::use_gas(&mut tx_ctx, 600).unwrap();
+        Core::use_gas(&mut tx_ctx, 600)
```
```suggestion
        Core::use_gas(&mut tx_ctx, 1)
```

pro-wh commented on runtime-sdk/src/types/transaction.rs:
```diff
@@ -148,6 +148,9 @@ pub enum CallResult {
 
         #[serde(rename = "code")]
```
why are these renamed?

kostko replied:
To make it explicit how the fields are called when serialized.

pro-wh replied:
hm ok

pro-wh commented:
update: there's now a draft of estimate gas. this version puts only the method registry in the context.

kostko commented on client-sdk/go/types/transaction.go:
```diff
@@ -184,8 +184,9 @@ func (cr *CallResult) IsSuccess() bool {
-	Code   uint32 `json:"code"`
+	Module  string `json:"module"`
+	Code    uint32 `json:"code"`
+	Message string `json:"message"`
```
The reason why there was no message field before was to reduce the output size. I guess we could at least make it optional (`omitempty` here).

pro-wh replied:
ok

kostko commented on runtime-sdk/src/types/transaction.rs:
```diff
@@ -148,6 +148,9 @@ pub enum CallResult {
         code: u32,
+
+        #[serde(rename = "message")]
+        message: String,
```
Make it optional using `#[serde(skip_serializing_if = "String::is_empty")]` and `#[serde(default)]`.

pro-wh replied:
ok

pro-wh replied:
I should say: these often won't be empty in our usage, with (all?) errors in this project having non-empty displays

kostko commented on runtime-sdk/src/modules/core/mod.rs:
```diff
@@ -30,10 +42,103 @@ pub enum Error {
+            .methods
+            .lookup_callable(&args.call.method)
+            .ok_or(Error::InvalidMethod)?;
+        let mut sim_ctx = DispatchContext {
+            mode: Mode::SimulateTx,
+            runtime_header: ctx.runtime_header,
+            runtime_round_results: ctx.runtime_round_results,
+            runtime_storage: ctx.runtime_storage,
+            io_ctx: ctx.io_ctx.clone(),
+            methods: ctx.methods,
+            max_messages: ctx.max_messages,
+            messages: Vec::new(),
+            values: BTreeMap::new(),
+        };
```
This should probably be factored out into a `with_simulation` method on the context.

pro-wh replied:
ok

kostko commented:
Nice!

> there's no batch-wide gas limit

I think we should support this. Would simply moving the gas accumulator from a per-tx value to a per-context value do the trick? I see that currently you are checking the tx gas limit directly in `use_gas` but maybe you could make the gas accumulator a struct with attributes like `batch_gas_limit` and `tx_gas_limit` (besides the `gas_used`) and then set the `tx_gas_limit` in the `AuthHandler` and `batch_gas_limit` in `BlockHandler`?

> adding message to CallResult::Failed. why did we discard this before? this will be more important now that modules will have to wrap errors from the core module for gas errors

Yeah the reason was to save on space.

I don't think that adding messages will solve the problem of wrapped errors as this still has issues if you want to match errors in a client application. I believe that the SDK should use [`source()`](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) on the errors to resolve the most inner SDK error and use the module/code from that error.

kostko commented:
> I don't think that adding messages will solve the problem of wrapped errors as this still has issues if you want to match errors in a client application. I believe that the SDK should use [`source()`](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) on the errors to resolve the most inner SDK error and use the module/code from that error.

See #100 for my attempt at implementing this.

pro-wh commented:
> > I don't think that adding messages will solve the problem of wrapped errors as this still has issues if you want to match errors in a client application. I believe that the SDK should use [`source()`](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) on the errors to resolve the most inner SDK error and use the module/code from that error.
> 
> See #100 for my attempt at implementing this.

cool, that should help too. 

> I don't think that adding messages will solve the problem of wrapped errors as this still has issues if you want to match errors in a client application. I believe that the SDK should use [`source()`](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) on the errors to resolve the most inner SDK error and use the module/code from that error.

Message and source seem to serve different purposes. The change in this branch to message is to make propagate it outside of the runtime, for example to help _debug_ when working with runtime clients. The source attribute seems to be meant for handling specific conditions from within the runtime, to help _develop_ certain functionality. I've looked at #100, and the intention there seems to be to strip most of the information before it leaves the runtime.

pro-wh commented:
> I think we should support this. Would simply moving the gas accumulator from a per-tx value to a per-context value do the trick?

_copying_ it maybe. we did want per-tx gas limit as well, judging from the auth info struct

> I see that currently you are checking the tx gas limit directly in `use_gas` but maybe you could make the gas accumulator a struct with attributes like `batch_gas_limit` and `tx_gas_limit` (besides the `gas_used`)

I'll see if we can do something that doesn't duplicate the data. 

> and then set the `tx_gas_limit` in the `AuthHandler` and `batch_gas_limit` in `BlockHandler`?

what should we set the batch gas limit from? maybe some runtime parameter?

pro-wh commented on runtime-sdk/src/storage/overlay.rs:
```diff
@@ -47,14 +47,17 @@ impl<S: Store> Store for OverlayStore<S> {
+    fn remove<K: AsRef<[u8]>>(&mut self, key: K) -> Option<Vec<u8>> {
         // For dirty values, remove from the overlay.
+        // TODO: Has this always been the case, where we diverge from what the Store trait defines
+        // as "remove" in the OverlayStore case?
```
I was going to see if it would be reasonable to make Store not throw away the last value in `insert` and `remove`, and I'm distracted by this `remove` implementation in OverlayStore. OverlayStore seems to change the meaning of `remove`. Here `remove` doesn't make future `get` return `None`. Here `remove` sort of resets it to the original value, which might still have been present.

kostko replied:
No, this does not reset it to the original value as it only removes it from the overlay, but not from the dirty set. So further `get` queries will still return `None`.

pro-wh replied:
oh right. None in overlay and key present in dirty means to remove on commit

kostko commented:
> _copying_ it maybe. we did want per-tx gas limit as well, judging from the auth info struct

No need to copy, as I mentioned we can have separate `batch_gas_limit` and `tx_gas_limit` in the accumulator and then the per-transaction limit can be updated in the auth handler. This works because we don't need to rollback gas accounting.

> what should we set the batch gas limit from? maybe some runtime parameter?

Yes it should be a runtime parameter that can be configured at genesis. Since you already made `core` a proper module, you could define some parameters and genesis for it.

kostko commented on runtime-sdk/src/modules/accounts/mod.rs:
```diff
@@ -75,12 +81,15 @@ pub enum Event {
 pub struct Parameters {
     #[serde(rename = "transfers_disabled")]
     pub transfers_disabled: bool,
+    #[serde(rename = "gas_cost_tx_transfer")]
+    pub gas_cost_tx_transfer: u64,
```
Maybe a separate "GasCosts" struct and then have "tx_transfer" as field?

pro-wh replied:
:shrug: okay

pro-wh commented:
incorporated changes from main. trying out rebasing this time, but it's a little weird with older non-transparent-error now appearing to come after the introduction of transparent errors. anyway.

also added batch gas. still keeping tx_values though. I think the explicitness compared to managing value lifecycles implicitly during authenticate_tx, not to mention doing non-authentication work \:

kostko commented on runtime-sdk/src/modules/accounts/mod.rs:
```diff
@@ -43,6 +45,14 @@ pub enum Error {
+    #[error("core: {0}")]
+    #[sdk_error(transparent)]
+    Core(
+        #[from]
+        #[source]
```
Since `from` implies `source` you only need `from`.

pro-wh replied:
thx I'll try that

pro-wh replied:
looks good

kostko commented on runtime-sdk/src/modules/accounts/mod.rs:
```diff
@@ -69,18 +79,29 @@ pub enum Event {
 pub struct Parameters {
     #[serde(rename = "transfers_disabled")]
     pub transfers_disabled: bool,
+    #[serde(rename = "gas_cost")]
```
```suggestion
    #[serde(rename = "gas_costs")]
```

pro-wh replied:
ah yes, here and in the key value runtime

kostko reviewed (commented):
> also added batch gas. still keeping tx_values though. I think the explicitness compared to managing value lifecycles implicitly during authenticate_tx, not to mention doing non-authentication work :

Looks good :+1:

pro-wh commented on client-sdk/ts-web/rt/src/accounts.ts:
```diff
@@ -12,6 +12,7 @@ export const MODULE_NAME = 'accounts';
 export const ERR_INVALID_ARGUMENT_CODE = 1;
 export const ERR_INSUFFICIENT_BALANCE_CODE = 2;
 export const ERR_FORBIDDEN_CODE = 3;
+export const ERR_CORE_CODE = 4;
```
oops, ts part needs more updates

pro-wh commented:
something's wrong with the genesis an parameter initialization. it keeps seeing batch_gas=0. how do I set up the core module parameters?

pro-wh commented:
it's because I was doing a query before the first transaction. I just found out that we don't call init_or_migrate until the first batch comes in, so we won't have a batch gas limit transferred from the genesis parameters yet ):

kostko commented:
> it's because I was doing a query before the first transaction. I just found out that we don't call init_or_migrate until the first batch comes in, so we won't have a batch gas limit transferred from the genesis parameters yet ):

That's true. We could call this for queries as well and we could also make executor nodes always propose a block on genesis (so that any migration can happen as soon as possible).

kostko commented on runtime-sdk/src/modules/core/mod.rs:
```diff
@@ -30,10 +40,206 @@ pub enum Error {
+    pub parameters: Parameters,
+}
+
+impl Default for Genesis {
```
Feel free to derive `Default`, #92 changes a bunch of these to use default derivation instead of explicitly defining these.

pro-wh replied:
ok

pro-wh replied:
converted from #92

kostko commented on runtime-sdk/src/context.rs:
```diff
@@ -147,9 +156,31 @@ impl<'a> DispatchContext<'a> {
+            mode: Mode::SimulateTx,
+            runtime_header: self.runtime_header,
+            runtime_round_results: self.runtime_round_results,
+            runtime_storage: self.runtime_storage,
```
Note that this way anything in the simulation context could still overwrite storage if it commits something. I think it is enough to mention this in the method documentation.

pro-wh replied:
would be better to have an overlay

kostko replied:
Yeah it would be if you can make it so easily.

pro-wh replied:
oh right I can't. I just realized there's no overlay defined for wrapping the panicky kind of MKVS. comment it is!

pro-wh replied:
https://github.com/oasisprotocol/oasis-core/issues/3904

kostko commented on runtime-sdk/src/context.rs:
```diff
@@ -271,6 +305,31 @@ impl<'a, 'b> TxContext<'a, 'b> {
+    /// Takes a value associated with the transaction.
+    ///
+    /// The previous value is removed so subsequent fetches will return the default value.
+    pub fn take_tx_value<V>(&mut self, key: &'static str) -> Box<V>
```
No tests for this one?

pro-wh replied:
adding

kostko commented on runtime-sdk/src/modules/core/mod.rs:
```diff
@@ -30,10 +40,206 @@ pub enum Error {
+#[serde(deny_unknown_fields)]
+pub struct Parameters {
+    #[serde(rename = "batch_gas")]
+    pub batch_gas: u64,
```
Should this be `max_batch_gas`?

pro-wh replied:
I'm fine with max_batch_gas. that would have been my second choice after batch_gas (the latter for consistency with `gas` in fee)

kostko commented on runtime-sdk/src/modules/core/mod.rs:
```diff
@@ -30,10 +40,206 @@ pub enum Error {
+    type Error = std::convert::Infallible;
+}
+
+pub trait API {
+    fn use_gas(ctx: &mut TxContext<'_, '_>, gas: u64) -> Result<(), Error>;
+    fn batch_use_gas(ctx: &mut DispatchContext<'_>, gas: u64) -> Result<(), Error>;
+}
```
Add documentation comments.

pro-wh replied:
moving to new thread

kostko commented on runtime-sdk/src/modules/core/mod.rs:
```diff
@@ -30,10 +40,206 @@ pub enum Error {
+    BatchOutOfGas,
+}
+
+/// Parameters for the accounts module.
```
```suggestion
/// Parameters for the core module.
```

kostko commented on runtime-sdk/src/modules/core/mod.rs:
```diff
@@ -30,10 +40,206 @@ pub enum Error {
+    pub batch_gas: u64,
+}
+
+impl Default for Parameters {
```
Feel free to derive `Default` if there's nothing non-default.

kostko commented on runtime-sdk/src/modules/core/mod.rs:
```diff
@@ -30,10 +40,206 @@ pub enum Error {
+}
+
+impl Module {
+    fn query_estimate_gas(
```
Add documentation comment.

pro-wh replied:
bumping

pro-wh commented:
merged with the context changes. I haven't gone and made a struct SimulationContext that implements Context. it turned out to be simpler to make a with_tx_simulation function. we'll revisit this later if we ever need a non-transaction-specific simulation context.

I tried writing use_gas in a new way with the optional transaction info, but we can't access the auth info and the values at the same time. thanks, rust

kostko commented on runtime-sdk/src/modules/core/mod.rs:
```diff
@@ -42,6 +52,64 @@ pub enum Error {
+    pub max_batch_gas: u64,
+}
+
+impl Default for Parameters {
```
This was resolved but it still has the custom Default implementation? Also below.

pro-wh replied:
woa let me check

pro-wh replied:
oh oops I must have gotten confused with similar changes in the account module. of course other PRs would not have touched the new code in core

kostko commented on runtime-sdk/src/modules/core/mod.rs:
```diff
@@ -51,3 +119,135 @@ pub mod state {
+impl API for Module {
+    fn use_gas<C: Context>(ctx: &mut C, gas: u64) -> Result<(), Error> {
+        let new_gas_used = match (
+            // todo: oh shoot we can't borrow for both of these at the same time
```
Yeah you'll need to refactor this. Probably just do something like `ctx.tx_auth_info().map(|ai| ai.fee.gas)` to avoid holding a reference to `ctx` just for the gas limit specified in the transaction.

pro-wh replied:
yup

kostko commented on runtime-sdk/src/modules/core/mod.rs:
```diff
@@ -42,6 +52,50 @@ pub enum Error {
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct Parameters {
+    #[serde(rename = "batch_gas")]
```
```suggestion
    #[serde(rename = "max_batch_gas")]
```

pro-wh commented on runtime-sdk/src/modules/core/mod.rs:
```diff
@@ -42,6 +52,50 @@ pub enum Error {
+    type Error = std::convert::Infallible;
+}
+
+pub trait API {
+    fn use_gas<C: Context>(ctx: &mut C, gas: u64) -> Result<(), Error>;
+}
```
ah yeah thanks for resurfacing this. I'll close out the old one

pro-wh commented:
- **we now run migration for queries too**
- finished fixing up smaller conflicts with main
- added doc comments
- incorporated other smaller review comments
